src/SArray.jl

```diff
@@ -43,13 +43,12 @@ end
     end
 end
 
-@generated function (::Type{SArray{S}})(x::T) where {S <: Tuple, T <: Tuple}
-    return quote
-        @_inline_meta
-        SArray{S, $(promote_tuple_eltype(T)), $(tuple_length(S)), $(tuple_prod(S))}(x)
```

Comment:
@andyferris ah hah; here's where we call `promote_tuple_eltype` from within a generated function — looking through this PR I think we'd only need to change this one line (by removing the `$` character :-) ) to fix the original problem. Is that right @andreasnoack?

Regardless of that it's always nice to remove some generated functions so I'm generally positive about doing these changes as a cleanup. Provided the original reason for needing `promote_tuple_eltype` is gone.

Reply:
Just tried it and it doesn't seem to be sufficient to fix the issue.

Reply:
That's weird, I was sure this would fix it. Locally removing the `$` seems to work for me:

```
julia> struct MyFloat64 <: Real
                x::Float64
              end

julia> Base.promote_rule(::Type{Float64}, ::Type{MyFloat64}) = MyFloat64

julia> SArray{Tuple{2}}((2.0, MyFloat64(1.0)))
2-element SArray{Tuple{2},MyFloat64,1,2} with indices SOneTo(2):
 MyFloat64(2.0)
 MyFloat64(1.0)
```

(btw beware that you can't always rely on `Revise` to work with this stuff... changes may require restarting Julia and reloading StaticArrays)

Reply:
Good catch, @c42f.

I guess if we just change this one line, then it should unambiguously have no negative impact on generated code. If we merge the whole PR we should probably also stress test this with long `Union`-type static arrays, etc, first.

Reply:
Not sure what went wrong but indeed it seems to be sufficient to fix the issue. I've opened https://github.com/JuliaArrays/StaticArrays.jl/pull/671. I still think we should consider geetting rid of the generated function here but it's not urgent.